pkg/apiserver/clusterinfo/host.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusterinfo
+
+import (
+	"math"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/jinzhu/gorm"
+)
+
+type CPUUsage struct {
+	Idle   float64 `json:"idle"`
+	System float64 `json:"system"`
+}
+
+type Memory struct {
+	Used  int `json:"used"`
+	Total int `json:"total"`
+}
+
+type Partition struct {
+	Path   string `json:"path"`
+	FSType string `json:"fstype"`
+	Free   int    `json:"free"`
+	Total  int    `json:"total"`
+}
+
+type HostInfo struct {
+	IP          string `json:"ip"`
+	CPUCore     int    `json:"cpu_core,omitempty"`
+	*CPUUsage   `json:"cpu_usage,omitempty"`
+	*Memory     `json:"memory,omitempty"`
+	Partitions  []PartitionInstance `json:"partitions,omitempty"`
+	Unavailable bool                `json:"unavailable"`
+}
+
+type Instance struct {
+	Address    string `gorm:"column:INSTANCE" json:"address"`
+	ServerType string `gorm:"column:TYPE" json:"server_type"`
+}
+
+type PartitionInstance struct {
+	Partition `json:"partition"`
+	Instance  `json:"instance"`
+}
+
+func GetAllHostInfo(db *gorm.DB) ([]HostInfo, error) {
+	hostMap, err := loadHosts(db)
+	if err != nil {
+		return nil, err
+	}
+	memory, usages, err := queryClusterLoad(db)
+	if err != nil {
+		return nil, err
+	}
+	cores, hostPartitionMap, err := queryClusterHardware(db)
+	if err != nil {
+		return nil, err
+	}
+	dataDirMap, err := queryDeployInfo(db)
+	if err != nil {
+		return nil, err
+	}
+
+	infos := make([]HostInfo, 0)
+	for ip, instances := range hostMap {
+		var partitions = make([]PartitionInstance, 0)
+		for _, instance := range instances {
+			ip := parseIP(instance.Address)
+
+			partitionMap, ok := hostPartitionMap[ip]
+			if !ok {
+				continue
+			}
+
+			dataDir, ok := dataDirMap[instance.Address]
+			if !ok {
+				continue
+			}
+
+			partition := inferPartition(dataDir, partitionMap)
+
+			partitions = append(partitions, PartitionInstance{
+				Partition: partition,
+				Instance:  instance,
+			})
+		}
+
+		info := HostInfo{
+			IP:         ip,
+			CPUCore:    cores[ip],
+			CPUUsage:   usages[ip],
+			Memory:     memory[ip],
+			Partitions: partitions,
+		}
+		infos = append(infos, info)
+	}
+
+	return infos, nil
+}
+
+func splitPath(path string) []string {
+	if path == "" {
+		return []string{}
+	}
+	return strings.Split(path, string(filepath.Separator))
+}
+
+func inferPartition(dataDir string, diskMap PartitionMap) Partition {
```

Comment:
can we simply use prefix check? The longest prefix should be the match one

Reply:
These two methods are same. I'd prefer this method.

Reply:
I'm worried about the correctness. For example, what's the purpose of `filepath.Rel(disk.Path, dataDir)` considering that `dataDir` may not be a directory inside `disk.Path`?